src/openvslam/tracking_module.cc

```diff
@@ -321,6 +355,28 @@ bool tracking_module::initialize() {
 
 bool tracking_module::track_current_frame() {
     bool succeeded = false;
+
+    if (update_pose_is_requested()) {
+        // Force relocalization by pose
+        curr_frm_.set_cam_pose(get_requested_pose());
+
+        curr_frm_.compute_bow();
+        const auto candidates = map_db_->get_close_keyframes(get_requested_pose(),
+                                                             pose_reloc_distance_threshold_,
+                                                             cos_angle_reloc_threshold_);
+        if (!candidates.empty()) {
+            succeeded = relocalizer_.reloc_by_candidates(curr_frm_, candidates);
+            if (succeeded) {
+                last_reloc_frm_id_ = curr_frm_.id_;
+            }
+        }
+        else {
+            curr_frm_.cam_pose_cw_is_valid_ = false;
+        }
+        finish_update_pose_request();
```

Comment:
If request_update_pose is called between the time the pose is retrieved and the relocalization succeeds, the request will be ignored.
I think we need either of the following:

- Make the request fail if update_pose_is_requested_ is already true.
- Use mutex between request retrieval and state change (finish_update_pose_request).

Reply:
In terms of mutex between any access or change the request status, it's already done in [there](https://github.com/OpenVSLAM-Community/openvslam/pull/79/files#diff-9d7c6e92591d28cb8c87b3973729fddb3abd564e9feae8861851fd48dd31b2caR195-R215). So, there is only need to add checks and bool return status to `tracker::update_pose_request()` and `system::update_pose()`.